contracts/PartnerFund.sol

```diff
@@ -0,0 +1,291 @@
+        address wallet;
+
+        DepositInfo[] depositsEther;
+        mapping(address => DepositInfo[]) depositsToken;
+
+        // Active balance of ethers and tokens.
+        int256 activeEtherBalance;
+        mapping(address => int256) activeTokenBalance;
+
+        // Staged balance of ethers and tokens.
+        int256 stagedEtherBalance;
+        mapping(address => int256) stagedTokenBalance;
+
+        DepositHistory[] depositsHistory;
+    }
+
+    //
+    // Variables
+    // -----------------------------------------------------------------------------------------------------------------
+    mapping(bytes32 => PerWalletInfo) private walletInfoMap;
+    mapping(address => bytes32) private addressTagMap;
+
+    //
+    // Events
+    // -----------------------------------------------------------------------------------------------------------------
+    event ChangedFeeEvent(bytes32 tag, uint256 fee);
+    event ChangedWalletEvent(bytes32 tag, address oldWallet, address newWallet);
+    event DepositEvent(bytes32 tag, int256 amount, address token); //token==0 for ethers
+    event StageEvent(bytes32 tag, address wallet, int256 amount, address token); //token==0 for ethers
+    event WithdrawEvent(bytes32 tag, address wallet, int256 amount, address token); //token==0 for ethers
+
+    //
+    // Constructor
+    // -----------------------------------------------------------------------------------------------------------------
+    constructor(address owner) Ownable(owner) public {
+    }
+
+    //
+    // Partner relationship functions
+    // -----------------------------------------------------------------------------------------------------------------
+    function setPartnerFee(bytes32 tag, uint256 fee) public onlyOwner notNullTag(tag) {
+        require(fee > 0);
+
+        walletInfoMap[tag].fee = fee;
+
+        //raise event
+        emit ChangedFeeEvent(tag, fee);
+    }
+
+    function getPartnerFee(bytes32 tag) public view notNullTag(tag) returns (uint256) {
+        return walletInfoMap[tag].fee;
+    }
+
+    function setPartnerAddress(bytes32 tag, address newWallet) public onlyOwner notNullTag(tag) {
+        //we allow wallet==0 to unlink
+        address oldWallet = walletInfoMap[tag].wallet;
+
+        walletInfoMap[tag].wallet = newWallet;
+
+        if (oldWallet != address(0))
+            addressTagMap[oldWallet] = 0x0;
+        if (newWallet != address(0))
+            addressTagMap[newWallet] = tag;
+
+        //raise event
+        emit ChangedWalletEvent(tag, oldWallet, newWallet);
+    }
+
+    function getPartnerAddress(bytes32 tag) public view returns (address) {
+        return walletInfoMap[tag].wallet;
+    }
+
+    //
+    // Deposit functions
+    // -----------------------------------------------------------------------------------------------------------------
+    function() public payable {
+        revert();
+    }
+
+    function receiveEthers(bytes32 tag) public notNullTag(tag) payable {
```

Comment:
My idea here was that `tag` should be of `address` type as in that case this contract implements the interface of `Beneficiary` abstract contract, which is really one of the key purposes of this contract.